### ISSUE TITLE
@typescript-eslint/explicit-module-boundary-types does not honour class implements interface

### ISSUE THREAD

**Issue by inspiraller** [typescript-eslint/typescript-eslint#2399]
For example:
This should pass. It doesn't. It says Missing return type on function getMethod.
```typescript
interface Declared {
  getMethod: () => number;
}

class SomeClass implements Declared {
  getMethod() {
    return 8;
  }
}
```

**Comment by bradzacher** [comment]
A few reasons we do not and will not honour interfaces for this rule:
1) The interface is not an explicit type - it's an implicit one because it's inherited.
2) If the interface is declared in a separate file, there is no colocated type for the function
3) Interrogating what things the interface defines requires type information, which this rule does not consume.

**Comment by inspiraller** [comment]
I'm not sure I understand why your rule breaks a valid typescript functionality. Incidentally, the example is not in a separate file. its in the same file. The reason to support this is to help encourage developers to declare their types before the class for readability. At the very least you could provide an option to exclude the rule for 'implements'. That way other linting plugins may deal specifically with type checking.

**Comment by bradzacher** [comment]
> I'm not sure I understand why your rule breaks a valid typescript functionality

This entire rule is technically contravene to typescript functionality, as there's no reason to explicitly declare any return types anywhere.

> Incidentally, the example is not in a separate file. its in the same file.

Sure, but things can get very complicated very quickly with this sort of thing, and there is exactly one case that is supportable without type information: when the interface has no external dependencies and is in the same file as the class.

Supporting one case like this causes some unpredictability in the rule, as users don't understand why the linter reports an error when they move the interface outside the file. This causes engineers to distrust the linter, and harms the entire linting ecosystem.

It's better to be consistent in your lack of support than inconsistent in your support.

> The reason to support this is to help encourage developers to declare their types before the class for readability

This is not a common pattern from what I've seen. Most codebases do not forward declare interfaces for classes because of ideals like DRY (don't repeat yourself).

> At the very least you could provide an option to exclude the rule for 'implements'

Options bring in complexity and maintenance burden. All contributors and maintainers are volunteers, so we endeavour to keep maintenance burden as low as possible.

In this instance just ignoring all classes with an `implements` is a bad solution because it's common for a class to include more members than what are in the interface.

Our implementation will not support such an option.
We also will not consume type information as it severely limits the usability of the rule - large portions of our userbase does not use type aware linting, so that means they cannot use the rule. So bringing is a major breaking change to the rule, which is a large reason we aren't going to do it here.

Note that this is all open source code - if you would like to fork our implementation and consume type information to support `implements` for your use-case, then you are welcome to do so! We have intentionally permissive licencing to allow you to do what you need with the codebase.

**Comment by ljosberinn** [comment]
We just stumbled across a probably related issue:

```ts
export abstract class SomeClass<T> {
  abstract set value(element: T);
}
```

Can't possibly declare a return type here:

`A 'set' accessor cannot have a return type annotation.`

Just to be sure this isn't a bug, Is this also an intentional decision?

**Comment by bradzacher** [comment]
@ljosberinn no. that is entirely unrelated to this issue.
our lint rules do not intentionally force you to produce invalid code.
that is a bug. please file a new issue.
